devito/operator.py

```diff
@@ -150,62 +157,76 @@ def arguments(self, *args, **kwargs):
 
         # Replace the overridden values with the provided ones
         for argname in o_vals.keys():
-            if not arguments[argname].shape == o_vals[argname].shape:
-                raise InvalidArgument("Shapes must match")
-
             arguments[argname] = o_vals[argname]
 
         # Traverse positional args and infer loop sizes for open dimensions
-        f_args = [f for f in arguments.values() if isinstance(f, SymbolicData)]
-        for f in f_args:
-            arguments[f.name] = self._arg_data(f)
+        f_args = [(name, f) for name, f in arguments.items()
+                  if isinstance(f, SymbolicData)]
+        for fname, f in f_args:
+            arguments[fname] = self._arg_data(f)
             shape = self._arg_shape(f)
 
             # Ensure data dimensions match symbol dimensions
             for i, dim in enumerate(f.indices):
-                # Infer open loop limits
-                if dim.size is None:
-                    # First, try to find dim size in kwargs
-                    if dim.name in kwargs:
-                        dim_sizes[dim] = kwargs[dim.name]
-
-                    if dim in dim_sizes:
-                        # Ensure size matches previously defined size
-                        if not dim.is_Buffered:
-                            assert dim_sizes[dim] <= shape[i]
+                # We don't need to check sizes for buffered dimensions
+                # against data shapes, all we need is the size of the parent.
+                if dim.is_Buffered:
+                    continue
+
+                # Check data sizes for dimensions with a fixed size
+                if dim.size is not None:
+                    if not shape[i] <= dim.size:
+                        error('Size of data argument for %s is greater than the size '
+                              'of dimension %s: %d' % (fname, dim.name, dim.size))
+                        raise InvalidArgument('Wrong data shape encountered')
                     else:
-                        # Derive size from grid data shape and store
-                        dim_sizes[dim] = shape[i]
-                else:
-                    if not isinstance(dim, BufferedDimension):
-                        assert dim.size == shape[i]
+                        continue
```

Comment:
Because in this case the `dim` has already been resolved at the code level, ie. `dim` is not part of the function signature because we hard-coded values in the kernel. In that case all we care about is that the data we run over adheres to the shape we hard-coded into the kernel.